From 67cbe6051b54e87ecb24184d43a5fb6b32cac288 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 09 Feb 2021 18:54:08 +0200
Subject: [PATCH] [SW-20154] gaudi2: add provision for eROM recovery

Based on a fuse bit setting, Gaudi2 ROM is expected to
execute pcie recovery code from ROM/eROM. This patch adds
support to check the efuse and take appropriate action.

Change-Id: If30abdf05e61f150e6890254a0769f1f43675193
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 0f7f1fd..fccfe3c 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -166,7 +166,7 @@
 
 	/* if recovery flag is set, bringup PCIe in recovery mode */
 	if (arcboot_rom_global.recovery_flag || (override == 0))
-		pcie_recovery();
+		execute_pcie_recovery();
 }
 
 static uint8_t arcboot_rom_fw_is_not_validated(void)
@@ -340,7 +340,7 @@
 			if (arcboot_rom_bringup_pcie_images(
 			    ARCBOOT_SECONDARY) != ARCBOOT_BRINGUP_OK) {
 				arcboot_rom_validation_fail();
-				pcie_recovery();
+				execute_pcie_recovery();
 				if ((arcboot_rom_global.config_fuse &
 				     EFUSE_SYSCFG_BACKUP_DISABLE) == 0) {
 					arcboot_rom_load_recovery();
diff --git a/common/pcie.c b/common/pcie.c
index eb5e368..d23174c 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -219,3 +219,8 @@
 	     EFUSE_SYSCFG_DISABLE_ROM_PCIE_RECOVERY) == 0)
 		pcie_config();
 }
+
+__attribute__((weak))void execute_pcie_recovery(void)
+{
+	pcie_recovery();
+}
diff --git a/common/pcie.h b/common/pcie.h
index 040176e..7475675 100644
--- a/common/pcie.h
+++ b/common/pcie.h
@@ -22,5 +22,6 @@
 #define PCIE_PHY_PROTOCOL0_EXT_RX_ADAPT_SEL_G4_CONF	0xFF
 
 void pcie_recovery(void);
+void execute_pcie_recovery(void);
 
 #endif /* _PCIE_H_ */
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index af86bf7..5e48eaf 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -64,3 +64,18 @@
 {
 	/* TODO: No HW FSM bits in Gaudi2, wait for ORON to comeback */
 }
+
+void execute_pcie_recovery(void)
+{
+	/* if eFUSE == ROM_PCIE_Recovery, execute directly */
+	if (arcboot_rom_global.config_fuse &
+			EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_ROM) {
+		pcie_recovery();
+		return;
+	}
+
+	/* else load PCIE binary from eROM to establish the link */
+
+	/* Further, load the recovery bin from eROM */
+
+}
diff --git a/gaudi2_arc_gnu/dev_efuse.h b/gaudi2_arc_gnu/dev_efuse.h
index 056db38..4ff9982 100644
--- a/gaudi2_arc_gnu/dev_efuse.h
+++ b/gaudi2_arc_gnu/dev_efuse.h
@@ -13,9 +13,10 @@
 #define EFUSE_SYSCFG_REVOKATION_LOCATION		7
 #define EFUSE_SYSCFG_REVOKATION_LOCATION_MASK		0x3
 
-#define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI	BIT(8)
-#define EFUSE_SYSCFG_KEY_LOCATION_ROM		BIT(5)
-#define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI	BIT(8)
+#define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI		BIT(8)
+#define EFUSE_SYSCFG_KEY_LOCATION_ROM			BIT(5)
+#define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI		BIT(8)
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_ROM		BIT(13)
 
 #define EFUSE_SYSCFG_DEFAULT	(EFUSE_SYSCFG_SPI_CONTROLLER | \
 				 EFUSE_SYSCFG_ENABLE_VALIDATION_MASK | \
